Configure for unit tests to run on circleci

jzoldak opened the pull request:
Documentation here: https://openedx.atlassian.net/wiki/display/TE/Continuous+Integration+of+edx-platform+on+CircleCI

Thanks @tusbar for getting us going on this.
@benpatterson @jbarciauskas I squashed your commits to make rebasing easier - thanks for your help too.

tusbar commented:
This is awesome @jzoldak! Thank you so much for doing this. 

I would maybe rename the `circle.sh` script into `circle-ci.sh`

benpatterson commented on circle.yml:
```diff
@@ -0,0 +1,53 @@
+    # Install a version which falls within that range.
+    - pip install pbr==0.9.0
+    - pip install -r requirements/edx/base.txt
+    - pip install -r requirements/edx/post.txt
```
Did we get rid of post.txt? What will happen here if that file disappears?

jzoldak replied:
We did, but that got reverted. I'll put in some defensive code in case it goes away again.
Generally, if any of the specifically-mentioned files get renamed/removed then the build will fail. 

Note that also if another prerequisites file gets ADDED in the edx-platform repo, it will not be used for creating the cache. It will however get picked up when the paver command for running tests gets executed.

benpatterson commented on circle.yml:
```diff
@@ -0,0 +1,53 @@
+test:
+  override:
+    - ./scripts/all-tests.sh:
+        parallel: true
```
Maybe a comment about what the parallel setting does (even if it's already in the circle docs)?

benpatterson commented on circle.yml:
```diff
@@ -0,0 +1,53 @@
+    - "if [ $(find reports -type f | wc -l) -gt 0 ] ; then cp -r reports/. $CIRCLE_TEST_REPORTS/junit ; fi":
+        parallel: true
+
+    # Configure your COVERALLS_REPO_TOKEN as an
```
What's the impact of this not-being set? Should it be the first thing that's executed in this file?

benpatterson commented on scripts/circle.sh:
```diff
@@ -0,0 +1,51 @@
+		paver test_lib --extra_args="--with-flaky" --cov_args="-p"
+		;;
+
+    *)  # catch-all (no-op)
```
Error message?

jzoldak commented:
jenkins run lettuce
jenkins run python

benpatterson commented:
heads up @jbarciauskas

jzoldak commented:
@benpatterson This is ready for review for merging. I found and fixed some issues with the scripts not passing through the exit codes correctly, and also updated to get rid of distribute now that the setuptools switchover is in master. 
Successfully building both on jenkins (this PR) on the edx org repo and circle.ci on my fork (See this build for example: https://circleci.com/gh/jzoldak/edx-platform/65)

benpatterson commented on scripts/circle-ci-tests.sh:
```diff
@@ -0,0 +1,73 @@
+
+EXIT=0
+
+case $CIRCLE_NODE_INDEX in
```
nit: describe what this index var means?

benpatterson commented:
:+1: once that typo is fixed and tests pass on the various systems.

jzoldak commented:
Checks passed on both jenkins and circle. Circle is clocking in at just under 40 minutes for the long straw (all of the lms unit tests). Squashing and merging now.

benpatterson commented:
:fireworks:

tusbar commented:
:+1:

tusbar commented:
@jzoldak just one thing: what do you think about setting the `NO_PREREQ_INSTALL` env variable to true?

I think that the installing the prereqs from paver test is take unnecessary time, resolving the github dependencies for example.

jzoldak commented:
@tusbar I like that idea. The only danger that I can think of is that if the list of requirements files changes (e.g. another is added, the order is changed, etc), then we lose the safety net of the requirements being correctly installed by the paver command.

That said, I think we can reduce this risk with a comment in prereqs.py reminding people to keep circle.yml in step with any changes.

I'll put in a PR and tag you, and we can continue the conversation there.
